### PR TITLE
Feature/104 - Login/Logout functionality

### PR DESCRIPTION
**Pull request by hamster4n**

### CODE REVIEW

File: src/main/java/co/gojohnny/configs/SecurityConfig.java [hunk 1]
```diff
@@ -48,6 +57,11 @@ public void configure(AuthenticationManagerBuilder auth) throws Exception {
     }
 
     @Bean
+    public UserLogoutSuccessHandler userLogoutSuccessHandler(){
+        return new UserLogoutSuccessHandler();
+    }
```
Comment (hamster4n) [inline src/main/java/co/gojohnny/configs/SecurityConfig.java]: its not created and i dont now why

File: src/main/webapp/WEB-INF/templates/index.html [hunk 2]
```diff
@@ -7,11 +7,26 @@
     <link href="https://www.mapbox.com/base/latest/base.css?v1.0" rel="stylesheet">
     <link href="https://fonts.googleapis.com/css?family=Open+Sans" rel="stylesheet">
     <link href="/static/css/map.css" rel="stylesheet">
+    <link href="../../static/css/login.css" rel="stylesheet" type="text/css">
+    <!-- Bootstrap CSS -->
+    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta.2/css/bootstrap.min.css"
+          integrity="sha384-PsH8R72JQ3SOdhVi3uxftmaW6Vc51MKb0q5P2rRUpPvrszuE4W1povHYgTpBfshb" crossorigin="anonymous">
```
Comment (hamster4n) [inline src/main/webapp/WEB-INF/templates/index.html]: I want to do view of my div like bootstrap view (for example like forms in another our html)

File: src/main/webapp/static/js/login_security.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    };
+
+    function getUserLoggedName() {
+        $.ajax({
+            type: "GET",
+            url: "http://localhost:8080/api/users/getUserName",
+            data: "text",
+            response: "text",
+            success: function (userName) {
+                return userName; //лажа
```
Comment (hamster4n) [inline src/main/webapp/static/js/login_security.js]: in this place I do not get a username (((

File: src/main/java/co/gojohnny/configs/SecurityConfig.java [hunk 4]
```diff
@@ -40,6 +43,12 @@ protected void configure(HttpSecurity http) throws Exception {
                     .loginPage("/users/login")
                     .usernameParameter("email")
                     .passwordParameter("password");
+
+        http
+                .logout()
+                .logoutSuccessUrl("/users/login")
+                .logoutUrl("users/logout")
+                .logoutSuccessHandler(userLogoutSuccessHandler);
```
Comment (hamster4n) [inline src/main/java/co/gojohnny/configs/SecurityConfig.java]: If I realized userLogoutSuccessHandler do I need  a logotSuccessUrl?

File: src/main/webapp/WEB-INF/templates/index.html (hunk 2 above)
Comment (boy4uck) [reply src/main/webapp/WEB-INF/templates/index.html]: Please don't spend much time on front-end questions related to HTML, and CSS (design stuff). Try to implement the task with the simplest UI first

File: src/main/webapp/static/js/login_security.js (hunk 3 above)
Comment (boy4uck) [reply src/main/webapp/static/js/login_security.js]: Of course, the method `getLoggedUserName()` in the controller is not correct. Security context does not hold a `User` instance. It stores `UserDetails`. Therefore, this code `SecurityContextHolder.getContext().getAuthentication().getPrincipal();` that returns a `UserDetails` instance can not be casted to `User`

File: src/main/java/co/gojohnny/configs/SecurityConfig.java (hunk 4 above)
Comment (boy4uck) [reply src/main/java/co/gojohnny/configs/SecurityConfig.java]: You don't even need `userLogoutSuccessHandler`. 
`.logout() .logoutUrl("users/logout")`.logoutSuccessUrl("/users/login");` is sufficient to implement logout functionality. 

However, please see how all security configurations are implemented. If you want to add logout configs you should use methods `and()`, and then add methods described above.

File: src/main/java/co/gojohnny/configs/SecurityConfig.java (hunk 1 above)
Comment (boy4uck) [reply src/main/java/co/gojohnny/configs/SecurityConfig.java]: We don't need this handler. you can create a handler only if you are going to implement some specific login on the logout.

File: src/main/java/co/gojohnny/configs/SecurityConfig.java (hunk 4 above)
Comment (hamster4n) [reply src/main/java/co/gojohnny/configs/SecurityConfig.java]: I try this option early and catch 404. Does I need to create logout.html?

File: src/main/webapp/static/js/login_security.js (hunk 3 above)
Comment (hamster4n) [reply src/main/webapp/static/js/login_security.js]: Anyway I must to get someone wrong String but I didnt. So there is two problem: wrong function and wrong method.

File: src/main/webapp/WEB-INF/templates/index.html (hunk 2 above)
Comment (hamster4n) [reply src/main/webapp/WEB-INF/templates/index.html]: I am not concentrate on this feature and think that is not take a lot of time. And it intresting to me how I can do it.

File: src/main/java/co/gojohnny/configs/SecurityConfig.java [hunk 5]
```diff
@@ -42,13 +39,11 @@ protected void configure(HttpSecurity http) throws Exception {
...
-                    .passwordParameter("password");
-
-        http
+                    .passwordParameter("password")
+                .and()
                 .logout()
-                .logoutSuccessUrl("/users/login")
                 .logoutUrl("users/logout")
-                .logoutSuccessHandler(userLogoutSuccessHandler);
+                .logoutSuccessUrl("/users/login");
```
Comment (hamster4n) [inline src/main/java/co/gojohnny/configs/SecurityConfig.java]: here I catch 404 (

File: src/main/webapp/static/js/login_security.js (hunk 3 above)
Comment (boy4uck) [reply src/main/webapp/static/js/login_security.js]: > Which context hold a User instance? 
There is no context that holds a `User` instance. You can get `UserDetails` from the context. It contains a username parameter, that stores logged in user's `email`;

File: src/main/java/co/gojohnny/web/api/security/UserLogoutSuccessHandler.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+import org.springframework.security.core.Authentication;
+import org.springframework.security.web.authentication.logout.LogoutSuccessHandler;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class UserLogoutSuccessHandler implements LogoutSuccessHandler {
```
Comment (boy4uck) [inline src/main/java/co/gojohnny/web/api/security/UserLogoutSuccessHandler.java]: This one should be removed

File: src/main/java/co/gojohnny/configs/SecurityConfig.java [hunk 7]
```diff
@@ -31,7 +31,7 @@ public SecurityConfig(UserService userService) {
     protected void configure(HttpSecurity http) throws Exception {
         http
                 .authorizeRequests()
-                .antMatchers(HttpMethod.POST, "/api/users").permitAll()
+                .antMatchers(HttpMethod.POST, "/api/users", "/api/users/current").permitAll()
```
Comment (hamster4n) [inline src/main/java/co/gojohnny/configs/SecurityConfig.java]: have trouble in this place

Comment (boy4uck) [reply src/main/java/co/gojohnny/configs/SecurityConfig.java]: I see :)

Comment (boy4uck) [reply src/main/java/co/gojohnny/configs/SecurityConfig.java]: You should allow `GET` requests to be performed

File: src/main/java/co/gojohnny/web/api/controller/UserRestController.java [hunk 8]
```diff
@@ -49,4 +51,15 @@ private void encryptUserPassword (User user){
         String encryptedPassword = bCryptPasswordEncoder.encode(user.getPassword());
         user.setPassword(encryptedPassword);
     }
+
+    @GetMapping("/current")
+    public String getLoggedUserName(){
+        UserDetails user = (UserDetails) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
```
Comment (boy4uck) [inline src/main/java/co/gojohnny/web/api/controller/UserRestController.java]: It will not contain a `UserDetails` instance when user is not logged in

**Comment by boy4uck** [comment]
@hamster4n, there are some conflicts that should be resolved before merge

**Comment by hamster4n** [comment]
Can't understand why conflicts still exists (

**Pull request merged**